Use mappings version to retrieve system index mappings at creation time

williamrandolph opened the pull request:
When creating system indices, we need to use mappings versions instead of `Version`. Now that these version numbers are available in cluster state (#99307), we can use them in `TransportCreateIndexAction` and `AutoCreateAction`.

Now that this is possible, we can deprecate the old way of retrieving prior system index descriptors, but we'll need to use in the security plugin in order to close out https://github.com/elastic/elasticsearch/issues/98505.

williamrandolph committed:
Use mappings version instead of node version for index creation

williamrandolph committed:
Fix TransportCreateIndexActionTests

williamrandolph committed:
spotless

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

elasticsearchmachine commented:
Hi @williamrandolph, I've created a changelog YAML for you.

williamrandolph committed:
Update docs/changelog/99555.yaml

williamrandolph committed:
Merge branch 'main' into sysindex/index-creation-mappings-version

williamrandolph committed:
Fix compilation error

williamrandolph committed:
Avoid default locale for string formatting

williamrandolph committed:
Merge branch 'main' into sysindex/index-creation-mappings-version

williamrandolph committed:
Restore accidental import removal

thecoop commented on server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java:
```diff
@@ -283,7 +283,7 @@ ClusterState execute(
 
                     if (isManagedSystemIndex) {
                         final SystemIndexDescriptor descriptor = mainDescriptor.getDescriptorCompatibleWith(
-                            currentState.nodes().getSmallestNonClientNodeVersion()
```
Is there a logic change between the smallest _non-client_ node version, and the min system index mapping version across all nodes?

williamrandolph replied:
In practice, I doubt it, because only the security index uses a prior index descriptor, and it hasn't changed in a long time.

I was assuming that "non-client" was a vestige from back when we had a transport client that could show up in the cluster, but looking at how `getSmallestNonClientNodeVersion()` is created, it means data nodes or the master node. So I guess in a situation where we have a non-data ingest node or coordinating node that has an older system index mapping, but wouldn't hold data, the new code would be too conservative and create the system index with the old mappings, and update the mappings once all nodes in the cluster catch up.

I'll see if there's a simple way to preserve the current behavior.

williamrandolph replied:
I committed some code that preserves the existing logic, but as I think about it, I'm not sure it's the best idea. The case where this could make a difference is when there's some non-data, non-master node on version X, but the master node and data nodes are all on version X+1. Using a smallest non-client node version lets the cluster upgrade the mappings without waiting for the version X node to upgrade, but if I understand things correctly, nothing would prevent a data node of version X from joining the mixed cluster, while the non-data/master node is still around.

I think we might want to change the logic to use the min system index mapping version across all nodes.

thecoop replied:
Yeah, that's a good point. Using the version across all nodes seems like the most sensible option here.

williamrandolph replied:
Thanks. I've undone the change.

williamrandolph committed:
Merge branch 'main' into sysindex/index-creation-mappings-version

williamrandolph committed:
Respond to PR feedback and fix tests

williamrandolph committed:
Merge branch 'main' into sysindex/index-creation-mappings-version

williamrandolph committed:
Forbidden APIs annotation

williamrandolph committed:
Merge branch 'main' into sysindex/index-creation-mappings-version

williamrandolph committed:
Get min cluster descriptor among all nodes

Previously, we retrieved a descriptor that matched the minimum
node version for "non-client" nodes, which in practice meant
master-eligible and data nodes. We're changing that logic here
because nodes are allowed to join the cluster if they match
the minimum system index mappings for the cluster, regardless
of node type. Here, we make a more conservative check so that
we don't upgrade mappings past what any potential node could
handle.

thecoop reviewed (approved):
Looks sensible. Generally, you could use `Strings.format("...")` rather than `String.format(Locale.ROOT, "...")`

williamrandolph committed:
String.format -> Strings.format
